bson/src/main/org/bson/BSONCallbackAdapter.java

```diff
@@ -83,12 +86,21 @@ protected void doWriteEndArray() {
 
     @Override
     protected void doWriteBinaryData(final BsonBinary value) {
-        if (value.getType() == BsonBinarySubType.UUID_LEGACY.getValue()) {
-            bsonCallback.gotUUID(getName(),
-                    readLong(value.getData(), 0),
-                    readLong(value.getData(), 8));
+        UuidRepresentation defaultUuidRepresentation = getDefaultUuidRepresentation();
+        if (defaultUuidRepresentation == UuidRepresentation.STANDARD) {
+            if (value.getType() == BsonBinarySubType.UUID_LEGACY.getValue()) {
```

Comment:
I think this needs to be changed to:
```suggestion
            if (value.getType() == BsonBinarySubType.UUID_STANDARD.getValue()) {
                UUID uuid = UuidHelper.decodeBinaryToUuid(value.getData(), value.getType(), defaultUuidRepresentation);
      bsonCallback.gotUUID(getName(), uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
      } else {
          bsonCallback.gotBinary(getName(), value.getType(), value.getData());
```
Without this change, all subtypes other than 3 will go through the `decodeBinary` function and will throw an exception.

Reply:
I fixed the bug in a slightly different way, by refactoring the code to extract common logic in both branches, and introduce a new top-level branch for non-UUID case.

Reply:
Changes make sense to me.